Migrate to tbb::task_group

Dr15Jones opened the pull request:
#### PR description:

The tbb::task ABI has been deprecated by Intel's developers in favor of tbb::task_group.

#### PR validation:

The code compiles and all framework related unit tests succeed.

Dr15Jones committed:
Switch to tbb::task_group

Remove use of deprecated tbb::task API in favor of tbb::task_group.

Dr15Jones committed:
Fixed unit tests to properly wait for threads.

Dr15Jones committed:
Wait in Worker needs its own task_group

Dr15Jones committed:
Reuse a task_group in EventProcessor

Dr15Jones committed:
Removed WaitingThreadIntProducer

Will not work properly using a tbb::task_group and was not a good
example even when using a tbb::task.

Dr15Jones committed:
Fixed copy and assignment operations

Dr15Jones committed:
Added utility methods to WaitingTask*Holder

Dr15Jones committed:
Fix FWCore/Concurrency unit tests

Properly use tbb::task_group

Dr15Jones committed:
Properly handle EventSetup IOVs with tbb::task_group

Have to be certain that task_group associated with an 'end of IOV' task is also the one used to do any waiting when doing synchronous operations.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/20977

- This PR adds an extra 472KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/20977/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/20977/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

Dr15Jones committed:
Applied code-format

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/20979

- This PR adds an extra 320KB to repository

Dr15Jones commented:
please test

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

FWCore/Concurrency
FWCore/Framework
FWCore/Integration
FWCore/TestProcessor
IOPool/Input


@makortel, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@makortel, @wddgit this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented:
hold

additional changes are expected but a fuller test is useful.

cmsbuild commented:
Pull request has been put on hold by  @Dr15Jones
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

cmsbuild commented:
-1

**Failed Tests**:   Build  ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-422c34/12683/summary.html
**COMMIT**: db792e20f48b823ced180f83b26ff68e37080742
**CMSSW**: CMSSW_11_3_X_2021-02-03-0800/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/32804/12683/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-03-0800/src/GeneratorInterface/LHEInterface/plugins/LHEFilter.cc
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-03-0800/src/GeneratorInterface/LHEInterface/plugins/LHE2HepMCConverter.cc
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-03-0800/src/GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-03-0800/src/GeneratorInterface/LHEInterface/plugins/ExternalLHEAsciiDumper.cc
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-03-0800/src/GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc: In lambda function:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-03-0800/src/GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc:303:45: error: no matching function for call to 'make_functor_task(tbb::internal::allocate_root_proxy, ExternalLHEProducer::beginRunProduce(edm::Run&, const edm::EventSetup&)::<lambda()>::<lambda()>)'
  303 |                                            });
      |                                             ^
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-03-0800/src/FWCore/Framework/src/Worker.h:47,
                 from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-03-0800/src/FWCore/Framework/src/Factory.h:5,
                 from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-03-0800/src/FWCore/Framework/interface/MakerMacros.h:4,
</pre>

## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

Dr15Jones committed:
Use tbb::task_group in ExternalLHEProducer

This replaces the use of the deprecated tbb::task API.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/20992

- This PR adds an extra 328KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc  modified in PR(s): #32167

cmsbuild commented:
Pull request #32804 was updated. @SiewYan, @smuzaffar, @Dr15Jones, @mkirsano, @makortel, @cmsbuild, @GurpreetSinghChahal, @agrohsje, @alberto-sanchez can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
-1

**Failed Tests**:   Build  ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-422c34/12696/summary.html
**COMMIT**: ca9690d875df460ec69562642594ac1f5d64468b
**CMSSW**: CMSSW_11_3_X_2021-02-03-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/32804/12696/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>Entering library rule at Mixing/Base
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-03-1100/src/Mixing/Base/src/BMixingModule.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-03-1100/src/Mixing/Base/src/PileUp.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-03-1100/src/Mixing/Base/src/SecondaryEventProvider.cc
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-03-1100/src/Mixing/Base/src/SecondaryEventProvider.cc: In member function 'void edm::SecondaryEventProvider::beginRun(edm::RunPrincipal&, const edm::EventSetupImpl&, const edm::ModuleCallingContext*, edm::StreamContext&)':
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-03-1100/src/Mixing/Base/src/SecondaryEventProvider.cc:40:20: error: 'class edm::WorkerManager' has no member named 'processOneOccurrence'; did you mean 'processOneOccurrenceAsync'?
   40 |     workerManager_.processOneOccurrence<OccurrenceTraits<RunPrincipal, BranchActionGlobalBegin> >(
      |                    ^~~~~~~~~~~~~~~~~~~~
      |                    processOneOccurrenceAsync
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-03-1100/src/Mixing/Base/src/SecondaryEventProvider.cc:40:97: error: expected primary-expression before '>' token
   40 |     workerManager_.processOneOccurrence<OccurrenceTraits<RunPrincipal, BranchActionGlobalBegin> >(
</pre>

## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

Dr15Jones committed:
Removed unnecessary include

Dr15Jones committed:
Converted SecondaryEventProvider to use tbb::task_group

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/21003

- This PR adds an extra 328KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc  modified in PR(s): #32167

cmsbuild commented:
Pull request #32804 was updated. @SiewYan, @smuzaffar, @civanch, @Dr15Jones, @mkirsano, @makortel, @mdhildreth, @cmsbuild, @GurpreetSinghChahal, @agrohsje, @alberto-sanchez can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
-1

**Failed Tests**:   Build  ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-422c34/12712/summary.html
**COMMIT**: 1d2e61864be914badd0a9687b82db0e3db4e2b97
**CMSSW**: CMSSW_11_3_X_2021-02-04-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/32804/12712/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>Leaving library rule at HeterogeneousCore/CUDACore
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-04-1100/src/HeterogeneousCore/CUDACore/test/test_ScopedContext.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-04-1100/src/HeterogeneousCore/CUDACore/test/test_main.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-04-1100/src/HeterogeneousCore/CUDACore/test/test_ScopedContextKernels.cu
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-04-1100/src/HeterogeneousCore/CUDACore/test/test_ScopedContext.cc: In function 'void ____C_A_T_C_H____T_E_S_T____0()':
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-04-1100/src/HeterogeneousCore/CUDACore/test/test_ScopedContext.cc:79:101: error: no matching function for call to 'make_waiting_task(tbb::internal::allocate_root_proxy, ____C_A_T_C_H____T_E_S_T____0()::<lambda(const std::__exception_ptr::exception_ptr*)>)'
   79 |             edm::make_waiting_task(tbb::task::allocate_root(), [](std::exception_ptr const* iPtr) {})};
      |                                                                                                     ^
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-04-1100/src/HeterogeneousCore/CUDACore/test/test_ScopedContext.cc:4:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-04-1100/src/FWCore/Concurrency/interface/WaitingTask.h:101:26: note: candidate: 'template<class F> edm::FunctorWaitingTask<F>* edm::make_waiting_task(F)'
  101 |   FunctorWaitingTask<F>* make_waiting_task(F f) {
</pre>

## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

Dr15Jones committed:
Updated ScopedContext to new WaitingTaskWithArenaHolder interface

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/21009

- This PR adds an extra 340KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc  modified in PR(s): #32167

cmsbuild commented:
Pull request #32804 was updated. @SiewYan, @smuzaffar, @civanch, @Dr15Jones, @mkirsano, @makortel, @mdhildreth, @cmsbuild, @GurpreetSinghChahal, @agrohsje, @fwyzard, @alberto-sanchez can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
-1

**Failed Tests**:   Build  ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-422c34/12716/summary.html
**COMMIT**: dd1122474d282cb7993ab3cb537bcb73fe9ebc27
**CMSSW**: CMSSW_11_3_X_2021-02-04-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/32804/12716/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Compiling bigobj /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-04-1100/src/SimG4Core/Application/src/ExceptionHandler.cc
>> Compiling bigobj /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-04-1100/src/SimG4Core/Application/src/LowEnergyFastSimModel.cc
>> Compiling bigobj /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-04-1100/src/SimG4Core/Application/src/PrimaryTransformer.cc
>> Compiling bigobj /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-04-1100/src/SimG4Core/Application/src/RunAction.cc
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-04-1100/src/SimG4Core/Application/src/RunManagerMTWorker.cc: In static member function 'static void RunManagerMTWorker::resetTLS()':
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-04-1100/src/SimG4Core/Application/src/RunManagerMTWorker.cc:199:88: error: no matching function for call to 'make_functor_task(tbb::internal::allocate_root_proxy, RunManagerMTWorker::resetTLS()::<lambda()>)'
  199 |     auto task = edm::make_functor_task(tbb::task::allocate_root(), []() { resetTLS(); });
      |                                                                                        ^
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-04-1100/src/SimG4Core/Application/src/RunManagerMTWorker.cc:21:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-04-1100/src/FWCore/Concurrency/interface/FunctorTask.h:44:19: note: candidate: 'template<class F> edm::FunctorTask<F>* edm::make_functor_task(F)'
   44 |   FunctorTask<F>* make_functor_task(F f) {
</pre>

## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

Dr15Jones committed:
RunManagerMTWorker now uses tbb::task_arena

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/21030

- This PR adds an extra 352KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc  modified in PR(s): #32167

cmsbuild commented:
Pull request #32804 was updated. @SiewYan, @smuzaffar, @civanch, @Dr15Jones, @mkirsano, @makortel, @mdhildreth, @cmsbuild, @GurpreetSinghChahal, @agrohsje, @fwyzard, @alberto-sanchez can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
-1

**Failed Tests**:   Build  ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-422c34/12740/summary.html
**COMMIT**: 652628c2bd8fc05bf8ef56f91580ce8089950c89
**CMSSW**: CMSSW_11_3_X_2021-02-05-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/32804/12740/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-05-1100/src/HeterogeneousCore/CUDATest/plugins/TestCUDAProducerGPUKernel.cu
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-05-1100/src/HeterogeneousCore/CUDATest/plugins/TestCUDAProducerGPUEWTask.cc:14:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-05-1100/src/HeterogeneousCore/CUDACore/interface/ScopedContext.h: In instantiation of 'void cms::cuda::impl::ScopedContextHolderHelper::pushNextTask(F&&, const cms::cuda::ContextState*) [with F = TestCUDAProducerGPUEWTask::acquire(const edm::Event&, const edm::EventSetup&, edm::WaitingTaskWithArenaHolder)::<lambda(cms::cuda::ScopedContextTask)>]':
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-05-1100/src/HeterogeneousCore/CUDACore/interface/ScopedContext.h:129:9:   required from 'void cms::cuda::ScopedContextAcquire::pushNextTask(F&&) [with F = TestCUDAProducerGPUEWTask::acquire(const edm::Event&, const edm::EventSetup&, edm::WaitingTaskWithArenaHolder)::<lambda(cms::cuda::ScopedContextTask)>]'
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-05-1100/src/HeterogeneousCore/CUDATest/plugins/TestCUDAProducerGPUEWTask.cc:92:4:   required from here
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-02-05-1100/src/HeterogeneousCore/CUDACore/interface/ScopedContext.h:231:100: error: no matching function for call to 'make_waiting_task_with_holder(std::remove_reference<edm::WaitingTaskWithArenaHolder&>::type, cms::cuda::impl::ScopedContextHolderHelper::pushNextTask(F&&, const cms::cuda::ContextState*) [with F = TestCUDAProducerGPUEWTask::acquire(const edm::Event&, const edm::EventSetup&, edm::WaitingTaskWithArenaHolder)::<lambda(cms::cuda::ScopedContextTask)>]::<lambda(edm::WaitingTaskWithArenaHolder)>)'
  231 |         replaceWaitingTaskHolder(edm::WaitingTaskWithArenaHolder{edm::make_waiting_task_with_holder(
      |                                                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
  232 |             std::move(waitingTaskHolder_), [state, func = std::forward<F>(f)](edm::WaitingTaskWithArenaHolder h) {
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  233 |               func(ScopedContextTask{state, std::move(h)});
</pre>

## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

Dr15Jones committed:
Corrected change to ScopedContext

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/21036

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc  modified in PR(s): #32167

cmsbuild commented:
Pull request #32804 was updated. @SiewYan, @smuzaffar, @civanch, @Dr15Jones, @mkirsano, @makortel, @mdhildreth, @cmsbuild, @GurpreetSinghChahal, @agrohsje, @fwyzard, @alberto-sanchez can you please check and sign again.

Dr15Jones commented:
Please test

cmsbuild commented:
-1

**Failed Tests**:   ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-422c34/12750/summary.html
**COMMIT**: 9475a03ba668737748f88cd5e695246294b4bac7
**CMSSW**: CMSSW_11_3_X_2021-02-06-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/32804/12750/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

Dr15Jones committed:
Fix clang warnings

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/21050

- This PR adds an extra 28KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc  modified in PR(s): #32167

cmsbuild commented:
Pull request #32804 was updated. @SiewYan, @smuzaffar, @civanch, @Dr15Jones, @mkirsano, @makortel, @mdhildreth, @cmsbuild, @GurpreetSinghChahal, @agrohsje, @fwyzard, @alberto-sanchez can you please check and sign again.

Dr15Jones commented:
please test

Dr15Jones commented:
@smuzaffar it looks like something happened to the PR tests as they appear to have finished from the logs but no GitHub comment was generated.

smuzaffar commented:
@Dr15Jones , the addOn job ( https://cmssdt.cern.ch/jenkins/job/ib-run-pr-addon/582/) was timeout after 2 hours so it was not able to update the status. bot waits for all jobs to provide their statuses before adding the result summary comment. 

I have restart the addOn job now and with in 15-20 mins your should see the results

smuzaffar commented:
@Dr15Jones , looks like cmsRun is running very slow with this change. Addon tests, which normally finish in 20 mins, is taking over 2 hours and relVals tests (single thread), which normally takes less than 1h30 took over 4h30.

Dr15Jones commented:
>  looks like cmsRun is running very slow with this change. Addon tests, which normally finish in 20 mins, is taking over 2 hours and relVals tests (single thread), which normally takes less than 1h30 took over 4h30.

It is quite possible that I have caused a 'resource starvation' issue in which cmsRun is caught in an infinite loop. This was one reason why I wanted to do an initial pull request to test for problems on a scale not feasible in my small work area.

Dr15Jones commented:
@smuzaffar looks like it got stuck again. I was really counting on seeing the results to help me focus on what needs to be fixed.

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-422c34/12754/summary.html
**COMMIT**: 05661a55525b41fa3d892ed2f78f10665bd59c4a
**CMSSW**: CMSSW_11_3_X_2021-02-07-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/32804/12754/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test materialBudgetTrackerPlots had ERRORS
---> test materialBudgetHGCalPlots had ERRORS
</pre>

## RelVals

The relvals timed out after 4 hours.

smuzaffar commented:
@Dr15Jones  I forced set the status and results summary is availabl enow

Dr15Jones commented:
@smuzaffar Thanks!

Dr15Jones commented:
So I looked at the failures in the RelVal test of the pull request. Two of the workflows got in a state where they could not make forward progress. Both were nanoAOD steps running two threads. In both cases, the two threads were stuck in `PATTauDiscriminationByMVAIsolationRun2:patTauDiscriminationByIsolationMVArun2v1DBoldDMwLTraw` where it was trying to do a synchronous get from the EventSetup system. I'm working to see if I can reproduce the same failure in my working area.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/21269

- This PR adds an extra 340KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc  modified in PR(s): #32167

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/21269/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/21269/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

Dr15Jones committed:
Applied code-format

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/21274

- This PR adds an extra 348KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc  modified in PR(s): #32167

cmsbuild commented:
Pull request #32804 was updated. @SiewYan, @smuzaffar, @civanch, @Dr15Jones, @mkirsano, @makortel, @mdhildreth, @cmsbuild, @GurpreetSinghChahal, @agrohsje, @fwyzard, @alberto-sanchez can you please check and sign again.

Dr15Jones commented:
this was just a rebase onto the master. Further changes are coming.

Dr15Jones committed:
Removed unnecessary tbb::empty_task

Dr15Jones committed:
Use tbb preview tbb::task::suspend/resume API

Dr15Jones committed:
Use edm::syncWait function

Because of how tbb implements its preview feature, the header file
must be included before any tbb header files in a translation unit.
The file name is meant to reflect that.

Dr15Jones committed:
Remove the need for a FunctorTask

Converting to a non mutable lambda allows calling tbb::task_group::run
directly.

Dr15Jones commented:
All the expected changes have been made.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/21278

- This PR adds an extra 472KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc  modified in PR(s): #32167

Dr15Jones commented:
Please test

cmsbuild commented:
-1

**Failed Tests**:   ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-422c34/13107/summary.html
**COMMIT**: da317d6eace90633e70c1b4ef15d4a1f6acf131c
**CMSSW**: CMSSW_11_3_X_2021-02-25-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/32804/13107/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

Dr15Jones committed:
Set lambda capture instead of a local variable

This was flagged by clang.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/21279

- This PR adds an extra 364KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc  modified in PR(s): #32167

cmsbuild commented:
Pull request #32804 was updated. @SiewYan, @smuzaffar, @civanch, @Dr15Jones, @mkirsano, @makortel, @mdhildreth, @cmsbuild, @GurpreetSinghChahal, @agrohsje, @fwyzard, @alberto-sanchez can you please check and sign again.

Dr15Jones commented:
Please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests AddOn 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-422c34/13108/summary.html
**COMMIT**: 275782c1bbffb92099feccbaec037a9d72bc68d5
**CMSSW**: CMSSW_11_3_X_2021-02-25-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/32804/13108/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test test_MpsWorkFlow had ERRORS
---> test test_PrepareInputDb had ERRORS
---> test testAlignmentOfflineValidation had ERRORS
---> test EcalDAQ_O2O_test had ERRORS
and more ...
</pre>

## AddOn Tests

- [hlt_data_Fake2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-422c34/13108/addOnTests/hlt_data_Fake2)```cmsDriver.py RelVal -s L1REPACK:Full --data --scenario=pp -n 10 --conditions auto:run2_hlt_Fake2 --relval 9000,50 --datatier "RAW" --eventcontent RAW --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2016 --fileout file:RelVal_Raw_Fake2_DATA.root --filein /store/data/Run2016B/JetHT/RAW/v1/000/272/762/00000/C666CDE2-E013-E611-B15A-02163E011DBE.root : FAILED - time: date Fri Feb 26 09:12:29 2021-date Fri Feb 26 09:03:44 2021 s - exit: 35584```
- [hlt_mc_Fake1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-422c34/13108/addOnTests/hlt_mc_Fake1)```cmsDriver.py TTbar_Tauola_13TeV_TuneCUETP8M1_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=pp -n 10 --conditions auto:run2_mc_Fake1 --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_25ns --fileout file:RelVal_Raw_Fake1_MC.root : FAILED - time: date Fri Feb 26 09:14:00 2021-date Fri Feb 26 09:04:29 2021 s - exit: 35584```
- [hlt_data_Fake1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-422c34/13108/addOnTests/hlt_data_Fake1)```cmsDriver.py RelVal -s L1REPACK:GCTGT --data --scenario=pp -n 10 --conditions auto:run2_hlt_Fake1 --relval 9000,50 --datatier "RAW" --eventcontent RAW --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_25ns --fileout file:RelVal_Raw_Fake1_DATA.root --filein /store/data/Run2015D/MuonEG/RAW/v1/000/256/677/00000/80950A90-745D-E511-92FD-02163E011C5D.root : FAILED - time: date Fri Feb 26 09:12:33 2021-date Fri Feb 26 09:05:22 2021 s - exit: 35584```


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2750983
 - DQMHistoTests: Total failures: 12
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2750948
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 36 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 156 log files, 37 edm output root files, 37 DQM output files

Dr15Jones committed:
Remove TBB thread observer when InitRootHandler's is destroyed

This is removed potential issues with order of destruction of
global objects from different translation units.

Dr15Jones committed:
Avoid use of global tbb::task_arena

Need to avoid end of job destruction order problems. The task_arena
used for the EventSetup is now managed by EventSetupsController.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/21291

- This PR adds an extra 72KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc  modified in PR(s): #32167

cmsbuild commented:
Pull request #32804 was updated. @SiewYan, @smuzaffar, @civanch, @Dr15Jones, @mkirsano, @makortel, @mdhildreth, @cmsbuild, @GurpreetSinghChahal, @agrohsje, @fwyzard, @alberto-sanchez can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals-INPUT AddOn 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-422c34/13137/summary.html
**COMMIT**: 82ef6b0c7b118e8ac3beca2a23a31cab98efa894
**CMSSW**: CMSSW_11_3_X_2021-02-26-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/32804/13137/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test TestRunnerPhysicsToolsCondLiteIO had ERRORS
---> test TestCTPPSDirectProtonSimulation had ERRORS
---> test materialBudgetTrackerPlots had ERRORS
---> test materialBudgetHGCalPlots had ERRORS
and more ...
</pre>

## RelVals-INPUT

- 11725.0```DAS Error```
- 11925.0```DAS Error```


## AddOn Tests

- [hlt_data_HIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-422c34/13137/addOnTests/hlt_data_HIon)```cmsDriver.py RelVal -s L1REPACK:Full --data --scenario=pp -n 10 --conditions auto:run3_hlt_HIon --relval 9000,50 --datatier "RAW" --eventcontent RAW --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run3_pp_on_PbPb --fileout file:RelVal_Raw_HIon_DATA.root --filein /store/data/Run2018D/HIMinimumBias0/RAW/v1/000/325/112/00000/660F62BB-9932-D645-A4A4-0BBBDA3963E8.root --customise_commands='if hasattr(process,"simMuonGEMPadTask"): setattr(process,"simMuonGEMPadTask",cms.Task()); from FWCore.ParameterSet.MassReplace import massSearchReplaceAnyInputTag; massSearchReplaceAnyInputTag(process.SimL1Emulator,"rawDataCollector","rawDataRepacker",False,True)' : FAILED - time: date Sat Feb 27 20:28:59 2021-date Sat Feb 27 19:20:49 2021 s - exit: 35584```
- [hlt_mc_HIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-422c34/13137/addOnTests/hlt_mc_HIon)```cmsDriver.py TTbar_Tauola_13TeV_TuneCUETP8M1_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=pp -n 10 --conditions auto:run3_mc_HIon --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run3_pp_on_PbPb --fileout file:RelVal_Raw_HIon_MC.root : FAILED - time: date Sat Feb 27 20:29:40 2021-date Sat Feb 27 19:21:10 2021 s - exit: 35584```
- [hlt_mc_Fake1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-422c34/13137/addOnTests/hlt_mc_Fake1)```cmsDriver.py TTbar_Tauola_13TeV_TuneCUETP8M1_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=pp -n 10 --conditions auto:run2_mc_Fake1 --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_25ns --fileout file:RelVal_Raw_Fake1_MC.root : FAILED - time: date Sat Feb 27 20:31:02 2021-date Sat Feb 27 19:21:16 2021 s - exit: 35584```


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 5 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2750983
 - DQMHistoTests: Total failures: 4
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2750957
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 36 files compared)
 - Checked 156 log files, 37 edm output root files, 37 DQM output files

Dr15Jones committed:
Properly pass down EventSetupInpl in calls

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/21317

- This PR adds an extra 16KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc  modified in PR(s): #32167

cmsbuild commented:
Pull request #32804 was updated. @SiewYan, @smuzaffar, @civanch, @Dr15Jones, @mkirsano, @makortel, @mdhildreth, @cmsbuild, @GurpreetSinghChahal, @agrohsje, @fwyzard, @alberto-sanchez, @santocch can you please check and sign again.

Commit:
Make ThreadTracker local to the main arena, and delete in ~InitRootHandlers()

Dr15Jones committed:
Merge pull request #5 from dan131riley/thread-observer-segfault-fix

Make ThreadTracker local to the main arena, and delete in ~InitRootHandlers()

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/21320

- This PR adds an extra 28KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc  modified in PR(s): #32167

cmsbuild commented:
Pull request #32804 was updated. @SiewYan, @smuzaffar, @civanch, @Dr15Jones, @mkirsano, @makortel, @mdhildreth, @cmsbuild, @GurpreetSinghChahal, @agrohsje, @fwyzard, @alberto-sanchez, @santocch can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-422c34/13174/summary.html
**COMMIT**: 3e2b8f2677e5fb3fb20e9db6c17331a64be10b72
**CMSSW**: CMSSW_11_3_X_2021-03-01-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/32804/13174/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test materialBudgetTrackerPlots had ERRORS
---> test materialBudgetHGCalPlots had ERRORS
</pre>

## RelVals-INPUT

- 11725.0```DAS Error```
- 11925.0```DAS Error```


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2750983
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2750958
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 36 files compared)
 - Checked 156 log files, 37 edm output root files, 37 DQM output files

Dr15Jones committed:
Properly shutdown PythonEventProcessor

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/21331

- This PR adds an extra 12KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc  modified in PR(s): #32167

cmsbuild commented:
Pull request #32804 was updated. @SiewYan, @smuzaffar, @civanch, @Dr15Jones, @mkirsano, @makortel, @mdhildreth, @cmsbuild, @GurpreetSinghChahal, @agrohsje, @fwyzard, @alberto-sanchez, @santocch can you please check and sign again.

Dr15Jones commented:
unhold

in principal, all the fixes are in. Now to do the review comments.

Dr15Jones commented on FWCore/Concurrency/interface/TaskBase.h:
```diff
@@ -0,0 +1,65 @@
+    unsigned int decrement_ref_count() { return --m_refCount; }
+
+  private:
+    virtual void recycle() { delete this; }
```
As you found, I want the ability to reuse some of the Tasks (as I had found that to help throughput in the toy framework).

makortel replied:
Ok

Dr15Jones commented on FWCore/Concurrency/src/SerialTaskQueue.cc:
```diff
@@ -28,15 +28,40 @@ SerialTaskQueue::~SerialTaskQueue() {
+    });
+    do {
+      g.wait();
+    } while (not done.load());
```
It is my intention to do so as that wait actually doesn't work since there are no running tasks for the newly created group :(. This should be fixed as part of this pull request.

Dr15Jones committed:
Properly wait for all tasks to finish in SerialTaskQueue dtr

Dr15Jones commented on FWCore/Framework/src/Worker.h:
```diff
@@ -953,7 +963,7 @@ namespace edm {
           ~DestroyTask() {
             auto p = m_task.load();
             if (p) {
-              tbb::task::destroy(*p);
+              delete p;
             }
```
Actually, this should be` p->recycle()`

Dr15Jones committed:
Added comment on use and finished the 'rule of 5'

Dr15Jones committed:
Use TaskSentry and remove unnneeded return statements

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/21344

- This PR adds an extra 48KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Services/plugins/InitRootHandlers.cc  modified in PR(s): #33038
  - File  GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc  modified in PR(s): #32167

cmsbuild commented:
Pull request #32804 was updated. @SiewYan, @smuzaffar, @civanch, @Dr15Jones, @mkirsano, @makortel, @mdhildreth, @cmsbuild, @GurpreetSinghChahal, @agrohsje, @fwyzard, @alberto-sanchez, @santocch can you please check and sign again.

makortel commented on FWCore/Framework/src/Worker.h:
```diff
@@ -924,7 +930,6 @@ namespace edm {
                            ParentContext const& parentContext,
                            typename T::Context const* context) {
     if (not workerhelper::CallImpl<T>::wantsTransition(this)) {
-      return;
```
Should this `return` be kept?

Dr15Jones commented on FWCore/Framework/src/Worker.h:
```diff
@@ -528,12 +526,10 @@ namespace edm {
                          std::exception_ptr* ptr = nullptr;
                          worker->runAcquireAfterAsyncPrefetch(ptr, info, parentContext, holder);
                        });
-            return;
```
This needs to stay

Dr15Jones commented on FWCore/Framework/src/Worker.h:
```diff
@@ -1034,7 +1026,6 @@ namespace edm {
                                         ParentContext const& parentContext,
                                         typename T::Context const* context) {
     if (not workerhelper::CallImpl<T>::wantsTransition(this)) {
-      return;
```
This needs to stay

Dr15Jones commented on FWCore/Framework/src/Worker.h:
```diff
@@ -448,12 +448,10 @@ namespace edm {
             } else {
               queue.push(*m_group, std::move(f));
             }
-            return;
```
This needs to stay

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32804/21345

- This PR adds an extra 456KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc  modified in PR(s): #32167

cmsbuild commented:
Pull request #32804 was updated. @SiewYan, @smuzaffar, @civanch, @Dr15Jones, @mkirsano, @makortel, @mdhildreth, @cmsbuild, @GurpreetSinghChahal, @agrohsje, @fwyzard, @alberto-sanchez, @santocch can you please check and sign again.

Dr15Jones commented:
Please test

Dr15Jones commented on FWCore/Framework/src/SynchronousEventSetupsController.h:
```diff
@@ -0,0 +1,52 @@
+      ~SynchronousEventSetupsController();
+
+      SynchronousEventSetupsController(SynchronousEventSetupsController const&) = delete;
+      SynchronousEventSetupsController const& operator=(SynchronousEventSetupsController const&) = delete;
```
Done

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-422c34/13221/summary.html
**COMMIT**: 73ec200c3b7fa9f40cc1738d9a5aa6d14bd2e96e
**CMSSW**: CMSSW_11_3_X_2021-03-02-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/32804/13221/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 2865 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2750983
 - DQMHistoTests: Total failures: 64465
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2686496
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 36 files compared)
 - Checked 156 log files, 37 edm output root files, 37 DQM output files

Dr15Jones commented:
All the differences in the test appear to come from just one workflow: 11634.911. That appears to be the only DD4Hep workflow.

Dr15Jones commented on FWCore/Concurrency/interface/SerialTaskQueue.h:
```diff
@@ -172,29 +164,16 @@ namespace edm {
-    QueuedTask<T>* pTask{new (tbb::task::allocate_root()) QueuedTask<T>{iAction}};
-    pTask->setQueue(this);
+  void SerialTaskQueue::push(tbb::task_group& iGroup, const T& iAction) {
+    QueuedTask<T>* pTask{new QueuedTask<T>{iGroup, iAction}};
```
As it is an internal detail I'm not as concerned with dealing with bare pointers.

makortel replied:
Ok

Dr15Jones commented on FWCore/Framework/src/Worker.h:
```diff
@@ -437,19 +441,19 @@ namespace edm {
+              queue.push(*m_group, std::move(f));
             }
-            return nullptr;
+            return;
```
Removing this return would change the logic since line 451 would be executed.

makortel replied:
No, need to keep.

makortel commented on FWCore/Framework/src/Worker.h:
```diff
@@ -511,23 +516,24 @@ namespace edm {
+                         std::exception_ptr* ptr = nullptr;
+                         worker->runAcquireAfterAsyncPrefetch(ptr, info, parentContext, holder);
+                       });
+            return;
```
No, need to keep.

Dr15Jones replied:
Removing this return would change the logic as line 534 would be called.

makortel replied:
Ok

Dr15Jones commented on FWCore/Framework/src/Worker.h:
```diff
@@ -965,27 +975,27 @@ namespace edm {
 
```
Under no circumstance should the destructor of this class be called at the same time as release :). There is actually only 1 thread which will ever be using this class at any given time so I'm not sure why I went with a `std::atomic` in the first place.

makortel replied:
Ok

Dr15Jones commented on FWCore/TestProcessor/src/TestProcessor.cc:
```diff
@@ -645,34 +659,40 @@ namespace edm {
-          streamLoopWaitTask->wait_for_all();
-          if (streamLoopWaitTask->exceptionPtr() != nullptr) {
-            std::rethrow_exception(*(streamLoopWaitTask->exceptionPtr()));
+          do {
+            taskGroup_.wait();
+          } while (not streamLoopWaitTask.done());
+          if (streamLoopWaitTask.exceptionPtr() != nullptr) {
+            std::rethrow_exception(*(streamLoopWaitTask.exceptionPtr()));
           }
```
It never got so annoying that I felt the need, unlike the `syncWait` case :).

makortel replied:
Ok

makortel commented:
This PR is effectively ready, but should really be tested for several IB cycles before a pre-release is cut. @silviodonato @qliphy When are you planning to build pre4? (effectively if that is on this week, as in the schedule, it's probably better for this PR to be merged after pre4 is built)

cvuosalo commented:
@Dr15Jones DD4hep uses tbb. Perhaps the changes in this PR somehow caused a re-calculation of the geometry by DD4hep, which then produced statistical fluctuations in the results of simulation using DD4hep. The comparison differences for 11634.911 are not significant.

civanch commented:
@Dr15Jones , what will happens if CMSSW is using tbb::task_group and some external continue to use tbb::task (even if external not really used it but have some header)?

@cvuosalo , if for any reason there is a re-computation of the geometry it is a bug. How this may be identified?

makortel commented:
@civanch 

> what will happens if CMSSW is using tbb::task_group and some external continue to use tbb::task (even if external not really used it but have some header)?

With the current TBB that setup continues to work (currently `tbb::task_group` uses `tbb::task` internally).

But an external using the deprecated `tbb::task` API would prevent us updating TBB. Do we have such an external?

makortel commented:
+1

I see pre4 being built (https://github.com/cms-sw/cmssw/issues/33067)

civanch commented:
+1

santocch commented:
+1

qliphy commented:
ping @cms-sw/generators-l2

agrohsje commented:
+generators

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1

mrodozov commented:
> @Dr15Jones DD4hep uses tbb. Perhaps the changes in this PR somehow caused a re-calculation of the geometry by DD4hep, which then produced statistical fluctuations in the results of simulation using DD4hep. The comparison differences for 11634.911 are not significant.

DD4Hep is not using TBB since https://github.com/cms-sw/cmsdist/pull/6554 when we removed it

vgvassilev commented on FWCore/Concurrency/interface/include_first_syncWait.h:
```diff
@@ -0,0 +1,40 @@
+//
+//  Created by Chris Jones on 2/24/21.
+//
+#define TBB_PREVIEW_RESUMABLE_TASKS 1
```
Hi @Dr15Jones, we suspect this define is causing a recent CXXMODULES IB regression. Would it be possible to move this define (I saw another one in the cc file) to the build system and possibly add it to the rootcling invocation (if scram does not do that automatically).


cc: @davidlange6

Dr15Jones replied:
> in one of the headers including tbb/task_group.h but not consistently.

That was intentional. It is only turned on where the feature is needed. Again, that was the recommendation of the tbb developers.

smuzaffar replied:
@Dr15Jones , yes we can set it at build time. @davidlange6 , are you testing it by adding it in the tbb.xml toolfile? You can add it in the toolfile and then scram should set itfo rboth normal build and rootcling

smuzaffar replied:
this macro is now defined via tbb toolfile ( see https://github.com/cms-sw/cmssw/pull/33140 and https://github.com/cms-sw/cmsdist/pull/6713 )

vgvassilev replied:
> > in one of the headers including tbb/task_group.h but not consistently.
> 
> That was intentional. It is only turned on where the feature is needed. Again, that was the recommendation of the tbb developers.

I'd assume they did not know about the fact we use C++ modules in the dictionaries. In theory, we can still use the macro the way we do, and only pass the relevant `-D` to rootcling. Not sure if that's worth it.
